src/notification/NotificationController.ts

```diff
@@ -0,0 +1,266 @@
+};
+
+export type GetNotificationState = {
+  type: `${typeof name}:getState`;
+  handler: () => NotificationState;
+};
+
+export type ShowNotification = {
+  type: `${typeof name}:show`;
+  handler: NotificationController['show'];
+};
+
+export type DismissNotification = {
+  type: `${typeof name}:dismiss`;
+  handler: NotificationController['dismiss'];
+};
+
+export type MarkNotificationRead = {
+  type: `${typeof name}:markRead`;
+  handler: NotificationController['markRead'];
+};
+
+export type GetNotifications = {
+  type: `${typeof name}:getNotifications`;
+  handler: NotificationController['getNotifications'];
+};
+
+export type GetNotificationCount = {
+  type: `${typeof name}:getCount`;
+  handler: NotificationController['getCount'];
+};
+
+export type GetUnreadNotificationCount = {
+  type: `${typeof name}:getUnreadCount`;
+  handler: NotificationController['getUnreadCount'];
+};
+
+export type ControllerActions =
+  | GetNotificationState
+  | ShowNotification
+  | DismissNotification
+  | MarkNotificationRead
+  | GetNotifications
+  | GetNotificationCount
+  | GetUnreadNotificationCount;
+
+export type NotificationMessenger = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions,
+  NotificationStateChange,
+  never,
+  never
+>;
+
+const metadata = {
+  notifications: { persist: true, anonymous: false },
+};
+
+const defaultState = {
+  notifications: {},
+};
+
+/**
+ * Controller that handles showing notifications to the user and rate limiting origins
+ */
+export class NotificationController extends BaseController<
+  typeof name,
+  NotificationState,
+  NotificationMessenger
+> {
+  private showNativeNotification;
+
+  /**
+   * Creates a NotificationController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.showNativeNotification - Function that shows a native notification in the consumer
+   */
+  constructor({
+    messenger,
+    state,
+    showNativeNotification,
+  }: {
+    messenger: NotificationMessenger;
+    state?: Partial<NotificationState>;
+    showNativeNotification: (origin: string, message: string) => void;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.showNativeNotification = showNativeNotification;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:show` as const,
+      (origin: string, args: NotificationArgs) => this.show(origin, args),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:dismiss` as const,
+      (ids: string[]) => this.dismiss(ids),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:markRead` as const,
+      (ids: string[]) => this.markRead(ids),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:getNotifications` as const,
```

Comment:
For the most part right now we don't use getter-style methods in the controllers. We use the controller state to communicate, and the application layer will consume the state directly and derive things from it and so on.

I guess snaps could benefit from these though :thinking: Is that what you had in mind - to expose these to snaps?

Reply:
I guess I was already partly thinking in selectors. I can remove the extra functions